src/DurableTask.AzureStorage/Partitioning/TablePartitionManager.cs

```diff
@@ -0,0 +1,843 @@
+                                partition.RowKey);
+                            
+                            response.WaitForPartition = true;
+                        }
+                        // Exception for failed table update due to ETag expired. 
+                        // Re-read the table immediately to get a new ETag.
+                        catch (RequestFailedException ex) when (ex.Status == 412)
+                        {
+                            this.settings.Logger.PartitionManagerInfo(
+                            this.storageAccountName,
+                            this.settings.TaskHubName,
+                            this.settings.WorkerId,
+                            partition.RowKey,
+                            $"Failed to update {partition.RowKey} due to concurrency. Re-read the table to get the latest Etag.");
+                            throw;
+                        }
+                        // Eat any exceptions during lease stealing.
+                        catch (Exception exception)
+                        {
+                            this.settings.Logger.PartitionManagerError(
+                                this.storageAccountName,
+                                this.settings.TaskHubName,
+                                this.settings.WorkerId,
+                                partition.RowKey,
+                                $"Unexpected error occurred in stealing {partition.RowKey} : {exception}"
+                                );
+                        }
+                    }
+                }
+            }
+
+            public void ClaimLease(TableLease lease)
+            {
+                lease.CurrentOwner = this.workerName;
+                lease.NextOwner = null;
+                lease.OwnedSince = DateTime.UtcNow;
+                lease.LastRenewal = DateTime.UtcNow;
+                lease.ExpiresAt = DateTime.UtcNow.AddMinutes(1);
+                lease.IsDraining = false;
+            }
+
+            public void DrainLease(TableLease lease, CloseReason reason)
+            {
+                lease.IsDraining = true;
+                var task = Task.Run(() => this.service.DrainTableLeaseAsync(lease, reason));
+                string partitionId = lease.RowKey!;
+                if (this.drainTasks.ContainsKey(partitionId))
+                {
+                    this.drainTasks[partitionId] = task;
+                }
+                else
+                {
+                    this.drainTasks.Add(partitionId, task);
+                }
+            }
+
+            public void ReleaseLease(TableLease lease)
+            {
+                lease.IsDraining = false;
+                lease.CurrentOwner = null;
+            }
+
+            public void RenewLease(TableLease lease)
+            {
+                lease.LastRenewal = DateTime.UtcNow;
+                lease.ExpiresAt = DateTime.UtcNow.Add(this.options.LeaseInterval);
+            }
```

Comment:
Can we change this to the following?
```suggestion
                Task drainTask = this.service.DrainTableLeaseAsync(lease, reason);
```
I'm concerned that wrapping it in `Task.Run(...)` makes the timing of the drain process less deterministic. Wrapping it in `Task.Run(...)` is also likely more expensive in terms of memory and CPU.